### PR TITLE
bootstrap: allow setting `--jobs` in config.toml

### LINKED ISSUE #131836: Bootstrap feature request: set `--jobs` in config.toml
To reduce RAM usage I'd like bootstrap to use fewer than all of my cores. (Many of them are efficiency cores so it's unclear that it's really so beneficial to use them all.) But currently I'd have to set `-j` on every single bootstrap invocation, which is not great. Would be nice if this could be set in the config file, like most other options can.

### PR DESCRIPTION
**Pull request by jieyouxu**
Allow setting `--jobs` in config.toml's `[build]` section.

```toml
[build]
jobs = 0
```

If this is unset or set to zero in config.toml, we look at `--jobs` flag. If that is also unset, then we fallback to `std::thread::available_parallelism`. If that is not available, then we default to `1`. The flags and `available_parallelism` fallback are already setup, this PR just adds a config.toml option to wire that up.

Closes #131836.

r? bootstrap

### CODE REVIEW

File: src/bootstrap/src/core/config/config.rs [hunk 1]
```diff
@@ -1621,6 +1623,12 @@ impl Config {
         // Verbose flag is a good default for `rust.verbose-tests`.
         config.verbose_tests = config.is_verbose();
 
+        if let Some(jobs) = jobs {
+            if jobs > 0 {
+                config.jobs = Some(jobs);
+            }
+        }
```
Comment (onur-ozkan) [inline src/bootstrap/src/core/config/config.rs:1630]: We should handle it here: https://github.com/rust-lang/rust/blob/e057ca519e5c557b8280efc7fca54d66199324df/src/bootstrap/src/core/config/config.rs#L1293

File: src/bootstrap/src/core/config/tests.rs [hunk 2]
```diff
@@ -352,3 +355,8 @@ fn parse_rust_std_features_empty() {
 fn parse_rust_std_features_invalid() {
     parse("rust.std-features = \"backtrace\"");
 }
+
+#[test]
+fn parse_jobs() {
+    assert_eq!(parse("build.jobs = 1").jobs, Some(1));
+}
```
Comment (onur-ozkan) [inline src/bootstrap/src/core/config/tests.rs:362]: We should add coverage for the case where flags override the configuration.

File: config.example.toml [hunk 3]
```diff
@@ -414,6 +414,10 @@
 # Specify the location of the Android NDK. Used when targeting Android.
 #android-ndk = "/path/to/android-ndk-r26d"
 
+# Number of parallel jobs to be used for building and testing. If set to `0` or
+# omitted, it will be automatically determined.
+#jobs = 0
```
Comment (onur-ozkan) [inline config.example.toml:419]: Worth to mention that this is used as --jobs flag on cargo.

File: src/bootstrap/src/core/config/config.rs (hunk 1 above)
Comment (jieyouxu) [reply src/bootstrap/src/core/config/config.rs]: Note that there's two layers of precedence, the line there is the top priority `--jobs` flag. I changed the logic to better reflect the precedence such that:

- `config.jobs` and `flags.jobs` are initially `None`.
- `config.jobs` is `Some` based on `--jobs=N` if `N > 0` and still `None` otherwise.
- If `config.jobs` is still `None`, then we fallback to `--set build.jobs` then `[build] jobs`. Note that the `build` toml section is read below this.

File: src/bootstrap/src/core/config/tests.rs (hunk 2 above)
Comment (jieyouxu) [reply src/bootstrap/src/core/config/tests.rs:360]: Added

File: config.example.toml (hunk 3 above)
Comment (jieyouxu) [reply config.example.toml]: Added a mention.

File: src/bootstrap/src/core/config/config.rs (hunk 1 above)
Comment (jieyouxu) [reply src/bootstrap/src/core/config/config.rs]: (Remark: I found this config layering somewhat hard to follow)

File: src/bootstrap/src/core/config/config.rs [hunk 4]
```diff
@@ -1511,8 +1513,19 @@ impl Config {
             metrics: _,
             android_ndk,
             optimized_compiler_builtins,
+            jobs,
         } = toml.build.unwrap_or_default();
 
+        if config.jobs.is_none() {
```
Comment (jieyouxu) [inline src/bootstrap/src/core/config/config.rs:1519]: Or, I could `match (flags.jobs, toml_job)` here too if that's easier to follow.

Comment (onur-ozkan) [reply src/bootstrap/src/core/config/config.rs:1519]: What about this:

```diff
diff --git a/src/bootstrap/src/core/config/config.rs b/src/bootstrap/src/core/config/config.rs
index abfc6178ce7..2f1334a9f38 100644
--- a/src/bootstrap/src/core/config/config.rs
+++ b/src/bootstrap/src/core/config/config.rs
@@ -1290,7 +1290,6 @@ pub(crate) fn parse_inner(
         config.rustc_error_format = flags.rustc_error_format;
         config.json_output = flags.json_output;
         config.on_fail = flags.on_fail;
-        config.jobs = flags.jobs.filter(|j| *j > 0);

         config.cmd = flags.cmd;
         config.incremental = flags.incremental;
@@ -1516,15 +1515,7 @@ fn get_table(option: &str) -> Result<TomlConfig, toml::de::Error> {
             jobs,
         } = toml.build.unwrap_or_default();

-        if config.jobs.is_none() {
-            config.jobs = match jobs {
-                Some(jobs) if jobs > 0 => Some(jobs),
-                _ => Some(
-                    std::thread::available_parallelism().map_or(1, std::num::NonZeroUsize::get)
-                        as u32,
-                ),
-            };
-        }
+        config.jobs = Some(threads_from_config(flags.jobs.unwrap_or(jobs.unwrap_or(0))));

         if let Some(file_build) = build {
```

Comment (jieyouxu) [reply src/bootstrap/src/core/config/config.rs:1519]: Works for me. I'll change it tmrw.

Comment (jieyouxu) [reply src/bootstrap/src/core/config/config.rs]: Modified to use this, gonna wait for CI to test it because the bootstrap tests hangs for me locally on msvc lol

File: src/bootstrap/src/core/config/config.rs [hunk 5]
```diff
@@ -1289,7 +1290,7 @@ impl Config {
         config.rustc_error_format = flags.rustc_error_format;
         config.json_output = flags.json_output;
         config.on_fail = flags.on_fail;
-        config.jobs = Some(threads_from_config(flags.jobs as u32));
+
```
Comment (onur-ozkan) [inline src/bootstrap/src/core/config/config.rs:1293]:
```suggestion
```

**Review by onur-ozkan — APPROVED** [review]
r=me with/without the nit suggestion

File: src/bootstrap/src/core/config/config.rs (hunk 5 above)
Comment (jieyouxu) [reply src/bootstrap/src/core/config/config.rs]: Oops left that yeah. Removed now.

**Comment by onur-ozkan** [comment]
Thanks!

@bors r+
